### PR TITLE
[Core] FormGroup component!

### PR DESCRIPTION
**Pull request by giladgray**
#### Changes proposed in this pull request:

add `FormGroup` component + tests + docs

#### Reviewers should focus on:

this component _does not support_ HTML props spread because it renders so many elements and we'd have several prop types conflicts (label, required, disabled).

took some creative liberties with the `required` implementation.

### CODE REVIEW

File: packages/core/src/components/forms/formGroup.tsx [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import * as classNames from "classnames";
+import * as PureRender from "pure-render-decorator";
+import * as React from "react";
+import * as Classes from "../../common/classes";
+import { IIntentProps, IProps } from "../../common/props";
+
+export interface IFormGroupProps extends IIntentProps, IProps {
+    /**
+     * Whether form group should appear as non-interactive.
+     * Remember that input elements must be disabled separately
```
Comment (cmslewis) [inline packages/core/src/components/forms/formGroup.tsx]: Missing period: `separately.`

File: packages/core/src/components/forms/formGroup.tsx [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+     */
+    disabled?: boolean;
+
+    /** Optional helper text. Will be wrapped in `.pt-form-helper-text` and displayed beneath `children`. */
```
Comment (cmslewis) [inline packages/core/src/components/forms/formGroup.tsx]: Can we reword this? Currently, the period happens to render on its own line, which is goofy:

image

File: packages/core/src/components/forms/formGroup.tsx [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    helperText?: React.ReactNode;
+
+    /** Whether to render label and control on a single line. */
```
Comment (cmslewis) [inline packages/core/src/components/forms/formGroup.tsx]: Ambiguous phrasing IMO: "control" looks like a verb here. Articles can fix this:

```
Whether to render the label and the control on a single line.
```

File: packages/core/src/components/forms/formGroup.tsx [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    inline?: boolean;
+
+    /** Label of this form group */
```
Comment (cmslewis) [inline packages/core/src/components/forms/formGroup.tsx]: Missing period: `group.`

Comment (cmslewis) [reply packages/core/src/components/forms/formGroup.tsx]: nit: Might also phrase as `label for...`, if it's not too confusing that we also have a `labelFor` prop. :)

File: packages/core/src/components/forms/formGroup.tsx [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+    /**
+     * `id` attribute of the labelable form element that this `FormGroup` controls,
+     * used as `<label for>` attribute.
+     */
+    labelFor?: string;
+
+    /**
+     * Whether this form input should appear as required (does not affect HTML form required status).
+     * Providing a boolean value will render a default "required" message after the `label` prop.
```
Comment (cmslewis) [inline packages/core/src/components/forms/formGroup.tsx]: nit: "a boolean `true` value"?

File: packages/core/src/components/forms/formGroup.tsx [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+     * Providing a JSX value will render that content instead.
+     *
+     * Note: the default message element is exposed as `FormGroup.DEFAULT_REQUIRED_CONTENT` and
```
Comment (cmslewis) [inline packages/core/src/components/forms/formGroup.tsx]: nit: consider emphasized styling for the word "Note"? e.g. "_Note:_" or "__Note:__"

File: packages/core/src/components/forms/formGroup.tsx [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+     *
+     * Defaults to `<span class="pt-text-muted">(required)</span>`.
+     */
+    public static DEFAULT_REQUIRED_CONTENT = <span className={Classes.TEXT_MUTED}>(required)</span>;
+
+    public render() {
+        const { children, helperText, label, labelFor } = this.props;
+        return (
+            <div className={this.getClassName()}>
+                <label className={Classes.LABEL} htmlFor={labelFor}>
```
Comment (cmslewis) [inline packages/core/src/components/forms/formGroup.tsx]: Whoa, why not just `for`?

File: packages/core/src/components/forms/formGroup.tsx [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+                    {label}
+                    {this.maybeRenderRequired()}
+                </label>
+                <div className={Classes.FORM_CONTENT}>
+                    {children}
+                    {helperText && <div className={Classes.FORM_HELPER_TEXT}>{helperText}</div>}
```
Comment (cmslewis) [inline packages/core/src/components/forms/formGroup.tsx]: nit: Looks this syntax messes with syntax highlighting in Github too. Note that `private`/`const`/`return` keywords below are mis-colored.

File: packages/core/src/components/forms/formGroup.tsx [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+                [Classes.DISABLED]: disabled,
+                [Classes.INLINE]: inline,
+            },
+            className,
+        );
+    }
+
+    private maybeRenderRequired() {
+        const { required } = this.props;
+        if (required === true) {
```
Comment (cmslewis) [inline packages/core/src/components/forms/formGroup.tsx]: Ternary?

```tsx
return required === true ? FormGroup.DEFAULT_REQUIRED_CONTENT : required;
```

File: packages/core/src/components/forms/formGroup.tsx (hunk 2 above)
Comment (giladgray) [reply packages/core/src/components/forms/formGroup.tsx]: ugh so lame! will do.

File: packages/core/src/components/forms/formGroup.tsx (hunk 7 above)
Comment (giladgray) [reply packages/core/src/components/forms/formGroup.tsx]: it's a React HTML thing, cuz `for` is a reserved word in JS! https://reactjs.org/docs/dom-elements.html#htmlfor

File: packages/core/src/components/forms/formGroup.tsx [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+    /**
+     * Whether this form input should appear as required (does not affect HTML form required status).
+     * Providing a boolean value will render a default "required" message after the `label` prop.
+     * Providing a JSX value will render that content instead.
+     *
+     * Note: the default message element is exposed as `FormGroup.DEFAULT_REQUIRED_CONTENT` and
+     * can be changed to provide a new global default for your app.
+     */
+    required?: boolean | React.ReactNode;
```
Comment (giladgray) [inline packages/core/src/components/forms/formGroup.tsx]: @cmslewis @llorca thoughts on calling this `requiredLabel` to clarify that it's cosmetic-only? that it has nothing to do with HTML's own `<input required>` attribute?

Comment (cmslewis) [reply packages/core/src/components/forms/formGroup.tsx]: I like that. 👍

File: packages/core/src/components/forms/formGroup.tsx (hunk 7 above)
Comment (cmslewis) [reply packages/core/src/components/forms/formGroup.tsx]: lol

File: packages/core/src/components/forms/formGroup.tsx (hunk 10 above)
Comment (llorca) [reply packages/core/src/components/forms/formGroup.tsx]: `requiredContent` to match the static default `FormGroup.DEFAULT_REQUIRED_CONTENT`?

File: packages/core/src/components/forms/formGroup.tsx [hunk 11]
```diff
@@ -18,7 +18,7 @@ export interface IFormGroupProps extends IIntentProps, IProps {
      */
     disabled?: boolean;
 
-    /** Optional helper text. Will be wrapped in `.pt-form-helper-text` and displayed beneath `children`. */
+    /** Optional helper text. The given content will be wrapped in `.pt-form-helper-text` and displayed beneath `children`. */
```
Comment (cmslewis) [inline packages/core/src/components/forms/formGroup.tsx]: lol

Comment (giladgray) [reply packages/core/src/components/forms/formGroup.tsx]: _effective_

File: packages/core/src/components/forms/formGroup.tsx [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+import * as classNames from "classnames";
+import * as PureRender from "pure-render-decorator";
+import * as React from "react";
+import * as Classes from "../../common/classes";
+import { IIntentProps, IProps } from "../../common/props";
+
+export interface IFormGroupProps extends IIntentProps, IProps {
+    /**
+     * Whether form group should appear as non-interactive.
+     * Remember that input elements must be disabled separately.
```
Comment (cmslewis) [inline packages/core/src/components/forms/formGroup.tsx]: Maybe include backticks: `input`?

File: packages/core/src/components/forms/formGroup.tsx [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+
+    /**
+     * Whether this form input should appear as required (does not affect HTML form required status).
+     * Providing a boolean `true` value will render a default "required" message after the `label` prop.
+     * Providing a JSX value will render that content instead.
+     *
+     * _Note:_ the default message element is exposed as `FormGroup.DEFAULT_REQUIRED_CONTENT` and
+     * can be changed to provide a new global default for your app.
+     */
+    requiredLabel?: boolean | React.ReactNode;
```
Comment (cmslewis) [inline packages/core/src/components/forms/formGroup.tsx]: This doesn't appear as `boolean | React.ReactNode` in the docs. It appears as simply `React.ReactNode`. Was confusing at a glance. Anything we need to do to fix that?

image

Comment (giladgray) [reply packages/core/src/components/forms/formGroup.tsx]: I believe that's because `ReactNode` is a union type that actually includes `boolean`.

Comment (giladgray) [reply packages/core/src/components/forms/formGroup.tsx]: the docs should be enough. oh but I should add an `@default`

**Pull request merged**